sunode/solver.py

```diff
@@ -252,12 +252,24 @@ def __init__(self, problem: Problem, *,
             sens_rhs = self._problem.make_sundials_sensitivity_rhs()
             self._init_sens(sens_rhs, sens_mode)
 
-    def _make_linsol(self) -> None:
-        linsolver = check(lib.SUNLinSol_Dense(self._state_buffer.c_ptr, self._jac))
-        check(lib.CVodeSetLinearSolver(self._ode, linsolver, self._jac))
-
-        self._jac_func = self._problem.make_sundials_jac_dense()
-        check(lib.CVodeSetJacFn(self._ode, self._jac_func.cffi))
+    def _make_linsol(self, linear_solver) -> None:
+        if linear_solver == "dense":
+            linsolver = check(lib.SUNLinSol_Dense(self._state_buffer.c_ptr, self._jac))
+            check(lib.CVodeSetLinearSolver(self._ode, linsolver, self._jac))
+
+            self._jac_func = self._problem.make_sundials_jac_dense()
+            check(lib.CVodeSetJacFn(self._ode, self._jac_func.cffi))
+        elif linear_solver == "dense_finitediff":
+            linsolver = check(lib.SUNLinSol_Dense(self._state_buffer.c_ptr, self._jac))
+            check(lib.CVodeSetLinearSolver(self._ode, linsolver, self._jac))
+        elif linear_solver == "spgmr":
+            linsolver = check(lib.SUNLinSol_SPGMR(self._state_buffer.c_ptr, lib.PREC_NONE, 5))
+            check(lib.CVodeSetLinearSolver(self._ode, linsolver, ffi.NULL))
+            check(lib.SUNLinSolInitialize_SPGMR(linsolver))
+            jac_prod = self._problem.make_sundials_jac_prod()
+            #check(lib.CVodeSetJacTimes(self._ode, ffi.NULL, jac_prod.cffi))
```

Comment:
I am not sure why I commented this check. This might indicate that something is not 100% right yet. :-/